File: services/control-plane-java/src/main/java/com/crossdesktopremote/controlplane/signaling/SignalingMessageLimits.java
package com.crossdesktopremote.controlplane.signaling;

/** Limit for a complete UTF-8 JSON envelope, including signed SDP metadata. */
final class SignalingMessageLimits {
	static final int MAX_TEXT_BYTES = 64 * 1024;
	static final String TOO_LARGE_REASON = "SIGNALING_MESSAGE_TOO_LARGE";

	private SignalingMessageLimits() {}
}

File: apps/client_flutter/lib/core/signaling/signaling_message_limits.dart
import 'dart:convert';

/// Budget for the complete JSON envelope, not the SDP or each WebSocket frame.
/// Keep in sync with Java SignalingMessageLimits. Signed content is never
/// truncated, recompressed or replaced by an unsigned offer to fit this limit.
abstract final class SignalingMessageLimits {
  static const maxTextBytes = 64 * 1024;
  static const closeReason = 'SIGNALING_MESSAGE_TOO_LARGE';
  static const userMessage = '信令消息超出接收限制，请更新并重启信令服务器及两端软件';

  static int validate(String payload) {
    final bytes = utf8.encode(payload).length;
    if (bytes > maxTextBytes) throw SignalingMessageTooLarge(bytes);
    return bytes;
  }
}

class SignalingMessageTooLarge implements Exception {
  const SignalingMessageTooLarge(this.payloadBytes);
  final int payloadBytes;

  @override
  String toString() =>
      '${SignalingMessageLimits.userMessage} '
      '($payloadBytes/${SignalingMessageLimits.maxTextBytes} bytes)';
}
